### PR TITLE
Checkpointing interval

### LINKED ISSUE #1264: Multiple undesired checkpoints created during single epoch
## 🐛 Bug

Thanks for the great project! When I sent custom `ModelCheckpoint` to `Trainer` and hoping to get one checkpoint each epoch, the Trainer eventually produced a lot of versioned checkpoints within a single epoch, wasting lots of disk space and were causing confusion. An example is shown as below:

```
checkpoints/
└── gan
    ├── _ckpt_epoch_0.ckpt
    ├── _ckpt_epoch_0_v0.ckpt
    ├── _ckpt_epoch_0_v10.ckpt
    ├── _ckpt_epoch_0_v11.ckpt
    ├── _ckpt_epoch_0_v12.ckpt
    ├── _ckpt_epoch_0_v13.ckpt
    ├── _ckpt_epoch_0_v14.ckpt
    ├── _ckpt_epoch_0_v15.ckpt
    ├── _ckpt_epoch_0_v16.ckpt
    ├── _ckpt_epoch_0_v17.ckpt
    ├── _ckpt_epoch_0_v18.ckpt
    ├── _ckpt_epoch_0_v19.ckpt
    ├── _ckpt_epoch_0_v1.ckpt
    ├── _ckpt_epoch_0_v20.ckpt
    ├── _ckpt_epoch_0_v21.ckpt
```

The minimal code I used for training (below I have showed how to reproduce this exactly):

```python
checkpoint_callback = ModelCheckpoint(
        filepath=os.path.join(
            "gan",
            experiment_name
        ),
        save_top_k=-1,
        period=10
    )
    
    trainer = Trainer(
        checkpoint_callback=checkpoint_callback,
		...
    )
```

### To Reproduce

1. Clone the repo

```bash
git clone https://github.com/leemengtaiwan/learnable_ai.git
git checkout 66b4cc9db5d1b9f1e2cf6c190d1172cdf6adbe92
```

2. Run the script to train a GAN with MNIST

```bash
cd learnable_ai/
python applications/image_generation/train_gan.py\
    --dataset mnist\
    --latent_dim 128\
    --dim 32\
    --channels 1\
    --batch_size 256 \
    --max_epochs 3
```

3. See the generated check points under `checkpoints`:

```bash
tree checkpoints/
```

4. There should be multiple checkpoints within a single epoch

```bash
checkpoints/
└── gan
    ├── _ckpt_epoch_0.ckpt
    ├── _ckpt_epoch_0_v0.ckpt
    ├── _ckpt_epoch_0_v10.ckpt
    ├── _ckpt_epoch_0_v11.ckpt
    ├── _ckpt_epoch_0_v12.ckpt
    ├── _ckpt_epoch_0_v13.ckpt
    ├── _ckpt_epoch_0_v14.ckpt
    ├── _ckpt_epoch_0_v15.ckpt
    ├── _ckpt_epoch_0_v16.ckpt
    ├── _ckpt_epoch_0_v17.ckpt
    ├── _ckpt_epoch_0_v18.ckpt
    ├── _ckpt_epoch_0_v19.ckpt
    ├── _ckpt_epoch_0_v1.ckpt
    ├── _ckpt_epoch_0_v20.ckpt
    ├── _ckpt_epoch_0_v21.ckpt
...
```

### Expected behavior

One checkpoint for each epoch when using custom `ModelCheckpoint` for `pl.LightningModule` which don't simply use `val_loss` for evaulation metric (thus the default checkpoint functionality on Trainer will not work)

### Environment

```
cuda:
        GPU:
                Tesla V100-SXM2-16GB
        available:           True
        version:             10.1
packages:
        numpy:               1.18.1
        pyTorch_debug:       False
        pyTorch_version:     1.4.0
        pytorch-lightning:   0.7.1
        tensorboard:         2.2.0
        tqdm:                4.43.0
system:
        OS:                  Linux
        architecture:
                64bit
                ELF
        processor:           x86_64
        python:              3.6.10
        version:             #1 SMP Tue Dec 24 03:25:32 UTC 2019
```

### Additional context

- [Related Slack discussion](https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1584788789243000)

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
Fixes #1264. In particular:
- explicitly count epochs
- move checkpointing in case missing valid to training end

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 1]
```diff
@@ -199,10 +199,9 @@ def set_distributed_mode(self, distributed_backend, num_gpu_nodes):
                 self.use_ddp2 = distributed_backend == 'ddp2'
 
             elif distributed_backend is None:
-                m = 'You requested multiple GPUs but did not specify a backend' \
-                    'Trainer(distributed_backend=dp) (or ddp, ddp2)' \
-                    'Setting distributed_backend=dp for you'
-                warnings.warn(m)
+                warnings.warn('You requested multiple GPUs but did not specify a backend'
+                              ' Trainer(distributed_backend=dp) (or ddp, ddp2)'
```
Comment (awaelchli) [inline pytorch_lightning/trainer/distrib_data_parallel.py]:
```suggestion
                              ' Trainer(distributed_backend=dp) (or ddp, ddp2).'
```

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 2]
```diff
@@ -199,10 +199,9 @@ def set_distributed_mode(self, distributed_backend, num_gpu_nodes):
...
+                              'Setting distributed_backend=dp for you')
```
Comment (awaelchli) [inline pytorch_lightning/trainer/distrib_data_parallel.py]:
```suggestion
                              ' Setting distributed_backend=dp for you.')
```

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -474,6 +473,12 @@ def run_training_epoch(self):
             if early_stop_epoch or self.fast_dev_run:
                 break
 
+        if not self.is_overriden('validation_step'):
+            self.call_checkpoint_callback()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py]: is it not a problem that this could get called twice now that you moved this down here? So basically the case "should_check_val=True" and "val_step not overridden". Is that fine?

Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: true, I ll fix it...

File: pytorch_lightning/profiler/profiler.py [hunk 4]
```diff
@@ -128,11 +128,12 @@ class AdvancedProfiler(BaseProfiler):
...
-            to std out when training is finished.
-        :param line_count_restriction (int|float): this can be used to limit the number of functions
-            reported for each action. either an integer (to select a count of lines),
-            or a decimal fraction between 0.0 and 1.0 inclusive (to select a percentage of lines)
+        Args:
+            output_filename: optionally save profile results to file instead of printing
+                to std out when training is finished.
+            line_count_restriction: this can be used to limit the number of functions
+                reported for each action. either an integer (to select a count of lines),
+                or a decimal fraction between 0.0 and 1.0 inclusive (to select a percentage of lines)
```
Comment (MattPainter01) [inline pytorch_lightning/profiler/profiler.py]: Why remove the type hints?

Comment (Borda) [reply pytorch_lightning/profiler/profiler.py]: I had them in args but it seems that disappeared while merging, I 'll fix it
